File: packages/economic_agents/src/economic_agents/interfaces/compute.py
"""Compute interface for managing agent compute resources."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime


@dataclass
class ComputeStatus:
    """Status of compute resources."""

    hours_remaining: float
    cost_per_hour: float
    balance: float
    expires_at: datetime
    status: str  # "active", "low", "expired"


class ComputeInterface(ABC):
    """Abstract interface for compute resource management."""

    @abstractmethod
    async def get_status(self) -> ComputeStatus:
        """Returns current compute status."""
        pass

    @abstractmethod
    async def add_funds(self, amount: float) -> bool:
        """Adds funds to compute account."""
        pass

    @abstractmethod
    async def get_cost_per_hour(self) -> float:
        """Returns current cost rate per hour."""
        pass

    @abstractmethod
    async def consume_time(self, hours: float) -> bool:
        """Consumes compute time, returns True if successful."""
        pass

File: packages/economic_agents/crates/economic-agents-interfaces/src/compute.rs
//! Compute resource interface.

use async_trait::async_trait;

use crate::{ComputeStatus, Currency, Hours, Result};

/// Interface for compute resource management.
///
/// Implementations may connect to real cloud providers (AWS, GCP),
/// mock compute tracking, or API-based compute services.
#[async_trait]
pub trait Compute: Send + Sync {
    /// Get current compute status.
    async fn get_status(&self) -> Result<ComputeStatus>;

    /// Add funds/hours to compute account.
    ///
    /// # Arguments
    /// * `amount` - Currency amount to add
    ///
    /// # Returns
    /// New compute status after funding.
    async fn add_funds(&self, amount: Currency) -> Result<ComputeStatus>;

    /// Consume compute time.
    ///
    /// # Arguments
    /// * `hours` - Hours to consume
    ///
    /// # Returns
    /// New compute status after consumption.
    async fn consume_time(&self, hours: Hours) -> Result<ComputeStatus>;

    /// Get the cost per hour.
    async fn get_cost_per_hour(&self) -> Result<Currency>;

    /// Get remaining hours.
    async fn get_hours_remaining(&self) -> Result<Hours>;

    /// Check if compute is available (hours > 0).
    async fn is_available(&self) -> Result<bool> {
        Ok(self.get_hours_remaining().await? > 0.0)
    }
}
